### PR TITLE
+htp Provide overloaded HttpApp.startServer method with system #1294

### PR DESCRIPTION
**Pull request by jlprat**
Issue: #1294
Add overloaded HttpApp.startServer method
Add docs for this new method
Add tests for this new method

### CODE REVIEW

**Comment by jlprat** [comment]
I was thinking to trim some of the current `startServer` methods, and just leave 1 or 2. Probably the one with `host`, `port` and maybe one with the `ActorSystem`.
The builder pattern, makes more and more sense now. And we can also offer `with` methods for the `ExceptionHandler` and `RejectionHandler` and similar.
I'll create an issue

**Comment by jrudolph** [comment]
Yep, let's solve this as another issue.

**Comment by jlprat** [comment]
Issue: https://github.com/akka/akka-http/issues/1300

File: akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala [hunk 1]
```diff
@@ -44,23 +44,31 @@ abstract class HttpApp extends Directives {
   }
 
   /**
-   * Start a server on the specified host and port, using provided settings.
+   * Start a server on the specified host and port, using the provided [[ActorSystem]].
+   * Note that this method is blocking
+   */
+  def startServer(host: String, port: Int, system: ActorSystem): Unit = {
+    startServer(host, port, ServerSettings(system), Some(system))
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala]: to be consistent with JavaDSL and also do what we want here, this should be `Option(system)` or we should explicitly validate it's not null

File: docs/src/test/scala/docs/http/scaladsl/HttpAppExampleSpec.scala [hunk 2]
```diff
@@ -48,11 +48,63 @@ class HttpAppExampleSpec extends WordSpec with Matchers
...
+        }
+    }
+
+    // Starting the server
+    val system = ActorSystem("ownActorSystem")
+    val settings = ServerSettings(ConfigFactory.load).withVerboseErrorMessages(true)
+    WebServer.startServer("localhost", 8080, settings, system)
+    system.terminate()
+    //#with-actor-system-settings
+  }
```
Comment (ktoso) [inline docs/src/test/scala/docs/http/scaladsl/HttpAppExampleSpec.scala]: this was just moving the test I think?

**Comment by ktoso** [comment]
Yeah, it's still not too crazy but I like the idea to explore the builder style here, could be good... :)

File: docs/src/test/scala/docs/http/scaladsl/HttpAppExampleSpec.scala (hunk 2 above)
Comment (jlprat) [inline docs/src/test/scala/docs/http/scaladsl/HttpAppExampleSpec.scala]: Yes, placing it next to the other ones

File: akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala (hunk 1 above)
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala]: To be fair we don't validate many fields for non-nullity, but sure, I can check if it's null and call further with `None`. This should do the trick

Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala]: Yup :) that works, thanks!

**Comment by ktoso** [comment]
Hmm I skimmed the other methods and they are a bit mixed in bevavoir. Javadsl will throw if null is passed in since optional.of does so, in Scala we have some Some and some Option which the first will lead to NPE later in an ugly way, and the later to silently falling back to creating a new system. Sorry to bother you with those but we should make them consistent - seems like throwing when we get a null explicitly Would be best IMHO so just like the javadsl does now

**Comment by jlprat** [comment]
We could use `Optional.ofNullable` in Java, this should behave like Scala's `Option.apply`.
I will go through other usages of Some, and convert them to the right version that supports null :)

**Comment by jlprat** [comment]
Amended the previous commit. Check again. Now all `Option` or `Optional` use the right version that supports null.s

**Comment by jlprat** [comment]
PLS BUILD

failure was https://github.com/akka/akka-http/issues/299

**Comment by jlprat** [comment]
PLS BUILD

Failure was https://github.com/akka/akka-http/issues/864

**Comment by ktoso** [comment]
I documented how the parameter is handled when null is passed in so people are not surprised, LGTM thanks!

**Comment by ktoso** [comment]
Grrrr... those trailing whitespace issues are driving me mad ;-)

**Comment by jlprat** [comment]
Do you want me to fix those trailing whitespaces?
Thanks for the extra doc in case of Null :)

**Comment by ktoso** [comment]
Probably easiest for you :) please pull, compile, push :)
Thanks and sorry for the hassle with this seemingly simple change :)

**Comment by ktoso** [comment]
(ok to merge once pushed and validated)

**Pull request merged**